middle_end/flambda2/simplify/simplify_binary_primitive.ml

```diff
@@ -1087,23 +1088,44 @@ let simplify_array_load (array_kind : P.Array_kind.t)
     let ty = T.bottom result_kind in
     let dacc = DA.add_variable dacc result_var ty in
     SPR.create_invalid dacc
-  | Ok array_kind ->
-    (* CR mshinwell: Add proper support for immutable arrays here (probably not
-       required at present since they only go into [Duplicate_array]
-       operations). *)
+  | Ok array_kind -> (
     let result_kind' =
       match accessor_width with
       | Scalar -> P.Array_kind.element_kind array_kind |> K.With_subkind.kind
       | Vec128 -> K.naked_vec128
     in
     assert (K.equal result_kind result_kind');
     let prim : P.t =
-      Binary (Array_load (array_kind, accessor_width, mutability), arg1, arg2)
+      Binary (Array_load (array_kind, accessor_width, mutability), array, index)
     in
-    let named = Named.create_prim prim dbg in
-    let ty = T.unknown (P.result_kind' prim) in
-    let dacc = DA.add_variable dacc result_var ty in
-    SPR.create named ~try_reify:false dacc
+    let[@inline] return_given_type ty ~try_reify =
+      let named = Named.create_prim prim dbg in
+      let dacc = DA.add_variable dacc result_var ty in
+      SPR.create named ~try_reify dacc
+    in
+    let[@inline] contents_unknown () =
+      return_given_type (T.unknown (P.result_kind' prim)) ~try_reify:false
+    in
+    match T.meet_is_immutable_array (DA.typing_env dacc) array_ty with
+    | Invalid -> SPR.create_invalid dacc
+    | Need_meet -> contents_unknown ()
+    | Known_result (_kind, fields, _mode) -> (
+      (* We try to read the array if it is known to be immutable, even if the
+         load is marked as mutable. *)
```

Comment:
I think this is correct, but I consider that using a mutable array load on an immutable array should be invalid. So checking the mutability of the load and only simplifying immutable loads should work in all safe cases and will be less surprising in unsafe cases (immutable arrays unsafely cast as mutable ones).